charms/metallb-controller/src/utils.py

```diff
@@ -202,6 +212,114 @@ def delete_namespaced_role_binding_with_api(name, namespace):
             )
 
 
+def get_k8s_version():
+    """Get k8s version with K8s API."""
+    logging.info("Getting k8s version with API")
+    _load_kube_config()
+
+    v1 = client.CoreV1Api()
+    try:
+        api_response = v1.list_node(pretty=True)
+        node = api_response.items[0]
+        kubelet_version = node.status.node_info.kubelet_version
+        version_tup = tuple(int(q) for q in re.findall("[0-9]+", kubelet_version)[:3])
+        return version_tup
+    except ApiException as e:
+        print("Exception when calling CoreV1Api->list_node: %s\n" % e)
+
+
+def get_pod_spec(image_info, cm):
+    """Get pod spec."""
+    version_tup = get_k8s_version()
+    rules = [
+        {
+            "apiGroups": [""],
+            "resources": ["services"],
+            "verbs": ["get", "list", "watch", "update"],
+        },
+        {
+            "apiGroups": [""],
+            "resources": ["services/status"],
+            "verbs": ["update"],
+        },
+        {
+            "apiGroups": [""],
+            "resources": ["events"],
+            "verbs": ["create", "patch"],
+        },
+        {
+            "apiGroups": [""],
+            "resources": ["nodes"],
+            "verbs": ["list"],
+        },
+    ]
+    if version_tup[:2] < (1, 25):
+        logging.info("Appending PSP-related podspec rules, kubelet version < 1.25.0")
+        rules.append(
+            {
+                "apiGroups": ["policy"],
+                "resourceNames": ["controller"],
+                "resources": ["podsecuritypolicies"],
+                "verbs": ["use"],
+            }
+        )
+    else:
+        logging.info("Skipping PSP-related podspec rules, kubelet version >= 1.25.0")
```

Comment:
If `version >= 1.25.0`, shouldn't we create a Pod Security Admission as in https://github.com/metallb/metallb/blob/main/config/manifests/metallb-native.yaml#L5?

Reply:
I don't think so. The namespace is created and managed by Juju, not the charm. I don't think it would be appropriate for the charm to alter security policy for a namespace/model that it does not fully own.

Most users won't encounter problems with this. Namespaces created by Juju do not have the pod-security annotations, so Kubernetes treats them as privileged by default.

If we do want to model Pod Security Admission in the Juju world, I think a model config option would be the most appropriate way to do so.